### ISSUE TITLE
Fix S6966 FP: Methods with different signatures should not trigger this rule.

### ISSUE THREAD

**Issue by ahusseini** [SonarSource/sonar-dotnet#9265]
### Description

S6966 is being triggered on methods that do not match in method signature. For example, in the MongoDB driver, `Find` and `FindAsync` are not related. `Find` returns `IFindFluent` for building chained mongo requests, while `FindAsync` returns `IAsyncCursor`. The following should not be triggering S6966.

`UnrelatedThing Find(string s, int i = 0)` or `UnrelatedThing Find(string s)`
`Task<string> FindAsync(string s)`

Only `string Find(string s)` should trigger this rule.

![CleanShot 2024-05-09 at 15 17 41@2x](https://github.com/SonarSource/sonar-dotnet/assets/32469060/7a457c00-d3c1-4c5e-ba1c-04bf579783bc)

### Repro steps

```csharp
_ = Find("123"); // INVALID
_ = await FindAsync("123"); // VALID
_ = FindSync("123"); // VALID
```

```csharp
private sealed record UnrelatedThing(string Name, int Number);

private UnrelatedThing Find(string s, int i = 0)
{
    return new UnrelatedThing(s, i);
}

private async Task<string> FindAsync(string s)
{
    return await Task.FromResult(s);
}

private string FindSync(string s)
{
    return s;
}
```

### Expected behavior

`_ = Find("123");` should not trigger S6966.

### Actual behavior

`_ = Find("123");` triggers S6966.

### Known workarounds

Please provide a description of any known workarounds.

### Related information

* C#/VB.NET Plugins version
* Visual Studio version
* MSBuild / dotnet version
* SonarScanner for .NET version (if used)
* Operating System

**Comment by MaStr11** [comment]
Thank you @ahusseini 

this is a known problem for Mongo DB's API design and has already been documented for the rule:

https://github.com/SonarSource/sonar-dotnet/blob/8b6f0f50f26d0280fbf115f1e5a18891b0a3417c/analyzers/tests/SonarAnalyzer.Test/TestCases/UseAwaitableMethod_MongoDBDriver.cs#L12-L21

In Mongo DB [FindSync](https://mongodb.github.io/mongo-csharp-driver/2.8/apidocs/html/M_MongoDB_Driver_IMongoCollection_1_FindSync__1.htm) is the synchronous version of [FindAsync](https://mongodb.github.io/mongo-csharp-driver/2.8/apidocs/html/M_MongoDB_Driver_IMongoCollection_1_FindAsync__1.htm) while [Find](https://mongodb.github.io/mongo-csharp-driver/2.8/apidocs/html/M_MongoDB_Driver_IMongoCollectionExtensions_Find__1.htm) is meant for something else. Because FindAsync and Find only differ in their return type, overload resolution is successful.

However, because the return types are not compatible, the overload should be rejected by our logic.

I marked this as a false positive and put it in our backlog.

**Comment by martin-strecker-sonarsource** [comment]
We have two options on how to solve this
1. We need to ensure that the (unwrapped) return type of the async method is covariant-compatible with the return type of the original method. This may produce FNs in a couple of cases like this
    * The return value is ignored on the call side
    * There is an explicit conversion of the return value to the type returned by the async version
    * `Nullable<struct>` vs `struct`
    * There might be all kinds of complicated cases with generics (e.g. [`IAdditionOperators`](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.iadditionoperators-3?view=net-8.0) vs `int`)
2. Exclude MongoDB's `Find` method.
